kernel/ipi.c

```diff
@@ -13,15 +13,62 @@ extern void z_trace_sched_ipi(void);
 #endif
 
 
-void flag_ipi(void)
+void flag_ipi(uint32_t ipi_mask)
 {
 #if defined(CONFIG_SCHED_IPI_SUPPORTED)
 	if (arch_num_cpus() > 1) {
-		_kernel.pending_ipi = true;
+		atomic_or(&_kernel.pending_ipi, (atomic_val_t)ipi_mask);
 	}
 #endif /* CONFIG_SCHED_IPI_SUPPORTED */
 }
 
+/* Create a bitmask of CPUs that need an IPI. Note: sched_spinlock is held. */
+atomic_val_t ipi_mask_create(struct k_thread *thread)
+{
+#if defined(CONFIG_IPI_OPTIMIZE)
+	uint32_t  ipi_mask = 0;
+	uint32_t  num_cpus = (uint32_t)arch_num_cpus();
+	uint32_t  id = _current_cpu->id;
+	struct k_thread *cpu_thread;
+
+	for (uint32_t i = 0; i < num_cpus; i++) {
+		if (id == i) {
+			continue;
+		}
+
+		/*
+		 * An IPI absolutely does not need to be sent if ...
+		 * 1. the CPU is not active, or
+		 * 2. <thread> can not execute on the target CPU
+		 * ... and might not need to be sent if ...
+		 * 3. the target CPU's active thread is not preemptible, or
+		 * 4. the target CPU's active thread has a higher priority
+		 *    (Items 3 & 4 may be overridden by a metaIRQ thread)
+		 */
+
+		cpu_thread = _kernel.cpus[i].current;
+		if ((cpu_thread != NULL) &&
+#if defined(CONFIG_SCHED_CPU_MASK)
+		    ((thread->base.cpu_mask & BIT(i)) != 0) &&
+#endif
+		    (thread_is_metairq(thread) ||
+		     (thread_is_preemptible(cpu_thread) &&
+		      (z_sched_prio_cmp(cpu_thread, thread) < 0)))) {
+			ipi_mask |= BIT(i);
+		}
+	}
+
+	return (atomic_val_t)ipi_mask;
+#elif (CONFIG_MP_MAX_NUM_CPUS > 1)
+	ARG_UNUSED(thread);
+
+	return IPI_ALL_CPUS_MASK;  /* Broadcast IPIs */
+#else
+	ARG_UNUSED(thread);
+
+	return 0;                                  /* No IPIs needed */
```

Comment:
Style nit: these two cases can be IMHO better expressed as simple if() expressions (and likely as early returns) than big #if blocks.  The compiler will remove the dead code if the kconfigs don't match, and I don't see any use of fields or APIs here that wouldn't work generically.  Zephyr is getting really bad about complicated preprocessor code blocks these days.